docs/design/slo_doc_tracking_for_slo_metric_data.md

```diff
@@ -0,0 +1,107 @@
+
+## Problem
+
+Today, SLO-metric-data stored in dashdotDB has no context of the SLO-document that the SLO-definition originated from. The [ServiceSLO table](https://github.com/app-sre/dashdotdb/blob/f88048cd9156afa93a4a0448ecd088c551c8408d/dashdotdb/models/dashdotdb.py#L168) does not contain any direct or indirect reference to SLO-documents.
+
+Consequently, if two (or more) SLO-definitions, spread across multiple SLO-documents, have identical names, service-ids, and namespace-ids, dashdotDB will not be able to store associated SLO-metric-data as desired. The SLO-metric-data values for these different SLO-definitions will effectively overwrite one another, and become indistinguishable when reading from the dashdotDB API.
+
+## Proposal
+The 'name' ([example](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L7)) of SLO docs will be introduced as an identifier for the SLO-metric-data stored in dashdotdb.
+
+The goals are:
+* be able to store SLO-doc-names alongside SLO-metric-data using dashdotDB's `POST /api/v1/serviceslometrics/{name}` HTTP API
+* be able to read SLO-doc-names alongside SLO-metric-data using dashdotDBs `GET /api/v1/serviceslometrics/metrics` and `GET /api/v1/serviceslometrics` HTTP APIs
+
+## Implementation - HTTP API
+
+---
+
+[`POST /api/v1/serviceslometrics/{name}`](https://github.com/app-sre/dashdotdb/blob/4663b71f34b2c540f7163985b7eccae8e7a9b0f0/dashdotdb/schemas/swagger.yaml#L241-L264)
+
+The JSON request body will now require a 'slo_doc.name' property.
+
+Example:
+```
+{
+    "name": "policy-engine-latency",
+    "SLIType": "latency",
+    "slo_doc": {
+        "name": "policy-engine"
+    },
+    "namespace": {
+        "name": "cincinnati-production"
+    },
+    "cluster": {
+        "name": "app-sre-prod-04"
+    },
+    "service": {
+        "name": "cincinnati"
+    },
+    "value": 66,
+    "target": 99
+}
+```
+
+---
+
+[`GET /api/v1/serviceslometrics/metrics`](https://github.com/app-sre/dashdotdb/blob/4663b71f34b2c540f7163985b7eccae8e7a9b0f0/dashdotdb/schemas/swagger.yaml#L193-L207)
+
+The objects within the JSON response body array will now include a 'slo_doc_name' property.
+
+Example: *(same as above)*
+
+---
+
+[`GET /api/v1/serviceslometrics`](https://github.com/app-sre/dashdotdb/blob/4663b71f34b2c540f7163985b7eccae8e7a9b0f0/dashdotdb/schemas/swagger.yaml#L208-L240)
+
+The JSON response body now includes a 'slo_doc_name' property.
+
+Example: *(same as above)*
+
+In addition, 'slo_doc_name' is added as an optional query parameter. The rationale for this being optional is to avoid breaking existing clients that are functioning fine without burdening them with needing to know a 'slo_doc_name'. A known risk with making this optional, however, is that clients potentially may make queries and get results from a 'slo_doc_name' that they don't intend to be querying from.
```

Comment:
I think it is worse to keep backwards compatibility but serve potentially incorrect data than to make a breaking change and ensure served data is correct.  We control everything that talks to DDDB atm so we can catch everything that has issues.  The intention is to correct the data problem, so imo we should accept that fixing the problem is a breaking change.

We had to do something similar when we introduced tokens.  We added non-optional fields because that was the only way to do things the right way.

Reply:
Sorry for coming late to the party, but I agree with @rrati. We control the clients for dashdotdb, so I would not make it optional and introduce a breaking change.

Reply:
Doc has been updated to reflect this as a required property.